File: src/pattern.rs
use crate::location::Located;

#[derive(Debug, Clone)]
pub(crate) enum Pattern<T> {
    Const(usize, Vec<T>),
    Tuple(Vec<T>),
    Binding(String),

    Char(char),
    Int(i64),
    Float(f64),
    Bool(bool),
}

#[derive(Debug, Clone)]
pub(crate) struct LocPattern<'a>(pub(crate) Located<'a, Pattern<LocPattern<'a>>>);

File: app/Pattern.hs
module Pattern ( Pattern(..), LocPatternT ) where

import Data.Text
import Literal
import Location
import Type

data Pattern a = Literal { ann :: a, l :: Literal }
               | Tuple { ann :: a, args :: [Pattern a] }
               | Binding { ann :: a, x :: Text }
               | Const { ann :: a, i :: Int, cargs :: [Pattern a] }
               | Or { ann :: a, lp :: Pattern a, rp :: Pattern a }
               | At { ann :: a, x :: Text, p :: Pattern a }

type LocPatternT = Pattern LocAnnT
